Support for OpenBSD

wezm opened the pull request:
Fixes #685 

These changes get nix building on OpenBSD 6.1. There is one failing test that I want a little guidance on:

```
error[E0308]: mismatched types
   --> src/sys/event.rs:333:30
    |
333 |     assert!(expected.data == actual.data());
    |                              ^^^^^^^^^^^^^ expected i64, found isize
```

`KEvent::data` is:

```
    pub fn data(&self) -> intptr_t {
        self.kevent.data as intptr_t
    }
```

I assume the test should be updated to cast to the expected type but wanted to confirm that before making the change.

Susurrus commented on src/sys/event.rs:
```diff
@@ -79,16 +78,18 @@ pub enum EventFilter {
         EV_CLEAR,
         EV_DELETE,
         EV_DISABLE,
+        #[cfg(not(target_os = "openbsd"))]
```
We prefer to whitelist versus blacklist, this appears to be on mac, ios, dragonly, freebsd and netbsd.

Susurrus replied:
Actually, OpenBSD does have it, see rust-lang/libc#613.

wezm replied:
By the looks of it they aren't in a released version of OpenBSD yet. The author of that patch added it to the current branch but there hasn't been a release since. I think the linked PR is still unmerged as a result.

I think the right thing to do for now is to leave it out. I can raise another issue to add it when the next OpenBSD release occurs.

Susurrus replied:
Sounds good. But please change this to a whitelist anyways. And add a `TODO` here that talks about adding it once OpenBSD is released (and name a specific version number so we don't need to look back in time to figure out this comment).

Susurrus commented on src/sys/signal.rs:
```diff
@@ -490,11 +490,13 @@ pub enum SigevNotify {
 
 /// Used to request asynchronous notification of the completion of certain
 /// events, such as POSIX AIO and timers.
+#[cfg(not(target_os = "openbsd"))]
```
Is this because libc doesn't support this on openbsd or is it because it's unsupported on OpenBSD?

wezm replied:
As far as I can tell it's not supported at all. No mention of sigevent in the man pages or [headers](https://github.com/openbsd/src/blob/941122a81b5d8c4d77e8a37913bd2e7d25e3de4b/sys/sys/signal.h).

Susurrus replied:
Might be nicer than to wrap this in a module and #cfg that module and then a single line importing everything from it instead of having all of these #cfgs.

Susurrus commented on src/sys/socket/consts.rs:
```diff
@@ -178,7 +178,10 @@ mod os {
+    #[cfg(not(target_os = "openbsd"))]
     pub const SO_TIMESTAMP: c_int           = libc::SO_TIMESTAMP;
+    #[cfg(target_os = "openbsd")]
+    pub const SO_TIMESTAMP: c_int           = 0x0800;
```
We don't hardcode values in `nix` anymore, this needs to be upstreamed into [libc](https://github.com/rust-lang/libc) before it can be included here.

Susurrus commented:
Regarding your specific question, yes it does appear that a cast is missing that'll use `as type_of_data`.

wezm commented:
Thanks for the quick review.

> Regarding your specific question, yes it does appear that a cast is missing that'll use as type_of_data.

So something like this?

```rust
assert!(expected.data as type_of_data == actual.data());
```

Susurrus commented:
I think the other way around. Look at the lines of code around it and it'll be similar.

asomers commented:
@wezm I would say that you should change the output type of the `data` function to `type_of_data`.

wezm commented:
@Susurrus @asomers Happy to fix either way. Changing the return type of `data` would be an API change. Just want to confirm that's ok before going down that route.

wezm commented:
Getting closer. All tests (excluding doc tests) are passing except this one:

```
---- sys::test_socket::test_scm_rights stdout ----
        thread 'sys::test_socket::test_scm_rights' panicked at 'Did not receive passed fd', /usr/obj/ports/rust-1.16.0/rustc-1.16.0-src/src/libcore/option.rs:715

test result: FAILED. 53 passed; 1 failed; 0 ignored; 0 measured
```

It's going to need bit of digging as the code looks right when compared to the example in [CMSG_DATA(3)](http://man.openbsd.org/CMSG_DATA) but there is a fair bit of unsafe work going on in `ControlMessage::ScmRights`.

Susurrus commented on test/test_unistd.rs:
```diff
@@ -242,7 +242,7 @@ fn test_fpathconf_limited() {
 fn test_pathconf_limited() {
     // AFAIK, PATH_MAX is limited on all platforms, so it makes a good test
-    let path_max = pathconf(".", PathconfVar::PATH_MAX);
+    let path_max = pathconf("/", PathconfVar::PATH_MAX);
```
Why the change here?

wezm replied:
See the commit message for that change. It was failing when the tests were run in parallel. Presumably one of them is doing something that makes . No longer exist.

asomers replied:
good catch.  There are indeed some tests that do that.

asomers commented on nix-test/src/const.c:
```diff
@@ -209,6 +209,31 @@ get_int_const(const char* err) {
     GET_CONST(ECANCELED);
 #endif
 
+#if defined(__OpenBSD__)
```
You can revert the changes to this file.  This file has been obsolete ever since most constants moved into libc.  libc has its own, superior version.

wezm replied:
I added these to get the errno tests passing. It looks like the [check_errno macro](https://github.com/wezm/nix/blob/11dae339cf8111ff326ecb3d6763c201e7aba757/src/errno.rs#L1920-L1924) uses them via [assert_const_eq](https://github.com/wezm/nix/blob/11dae339cf8111ff326ecb3d6763c201e7aba757/nix-test/src/lib.rs#L17-L27) or did I miss something?

asomers replied:
Grr, I thought we had already removed that.  Go ahead and leave these changes in for now; we'll delete them in a future PR.

Susurrus replied:
As we do conversions of our code to use upstream libc values, we can remove these tests incrementally until they're all gone. But for now, we should add these tests if we have any hard-coded values (like all the `errno` values currently are).

wezm commented:
> test_scm_rights failure on FreeBSD
Yep I'm working on this at the moment. It's proving quite a challenge to get all OSes working at once. Might take a little bit to get sorted but I am working on it a bit each day.

wezm commented:
It was an adventure but I think 403cc5e finally resolves the failing control message test (for all platforms). As far as I can tell it was a bug in the implementation that only exposed itself on OpenBSD. More details in the commit message. Next up I'll deal with the other requested changes and get the doc tests passing. Not sure what's up with AlanCI failing to find the commit... but the tests pass on my FreeBSD VM.

asomers commented:
Looks like the OSX build failure is trivial to fix.  And don't forget to remove the hardcoded `SO_TIMESTAMP` value.

wezm commented:
Ok folks, I think all requested changes have been addressed. I've replaced the SO_TIMESTAMP magic number with the one I upstreamed in libc. Should be good to merge now.

Susurrus commented on src/sys/signal.rs:
```diff
@@ -3,7 +3,9 @@
 use libc;
 use {Errno, Error, Result};
+#[cfg(not(target_os = "openbsd"))]
 use std::fmt;
```
These can be combined into one `use` statement like `use std::fmt::{self, Debug};`.

Susurrus replied:
This still isn't addressed.

Susurrus commented on src/sys/signal.rs:
```diff
@@ -488,102 +489,114 @@ pub enum SigevNotify {
-    sigevent: libc::sigevent
-}
+#[cfg(not(target_os = "openbsd"))]
+pub mod sigevent {
```
I don't think this module should be public since you glob-import everything into the `sys::signal` module above.

Susurrus commented:
You've got a couple of commits in there that I think should be merged with earlier ones as they're fixing problems that were created by this PR. You also have a merge commit there in the middle. I'd prefer if you rebased your changes rather then merged master back into your branch (makes it harder to read the history when there are a ton of merge commits on top of the ones `bors` adds).

This could also use a CHANGELOG entry if `0.9` didn't work on OpenBSD and the next release will.

I suggest you wait until #647 as it changes a bunch of socket constants.

wezm commented:
Righto, addressed the comments on `src/sys/signal.rs`, rebased against current master and added a CHANGELOG entry.

Susurrus commented on src/sys/socket/ffi.rs:
```diff
@@ -23,6 +23,14 @@ pub type type_of_cmsg_data = c_uint;
+
+#[cfg(not(target_os = "linux"))]
+pub type type_of_msg_iovlen = c_uint;
+
```
There should only be one empty line following this.

Susurrus commented on src/sys/socket/ffi.rs:
```diff
@@ -33,9 +41,9 @@ pub struct msghdr<'a> {
     pub msg_name: *const c_void,
```
This struct should be replaced with the FFI types declared in libc, but that doesn't need to be done for this PR. Please add a FIXME stating that.

Susurrus commented on src/sys/socket/mod.rs:
```diff
@@ -235,6 +235,7 @@ impl<'a> Iterator for CmsgIterator<'a> {
 
         // Advance our internal pointer.
         if cmsg_align(cmsg_len) > buf.len() {
+            println!("cmsg_align(cmsg_len) > buf.len(): {} > {}", cmsg_align(cmsg_len), buf.len());
```
I think this debugging statement should be removed, yes?

Susurrus commented on CHANGELOG.md:
```diff
@@ -12,6 +12,8 @@ This project adheres to [Semantic Versioning](http://semver.org/).
   ([#672](https://github.com/nix-rust/nix/pull/672))
 - Added protocol families in `AddressFamily` enum.
   ([#647](https://github.com/nix-rust/nix/pull/647))
+- Added support for OpenBSD.
```
We didn't "add support". OpenBSD cannot be tested in CI yet, so we haven't added support here, but this PR does do is fix compilation and tests for OpenBSD targets.

Susurrus commented:
Okay, my previous comments don't make sense if you look at the entire changelog because you have commits that fix issues that you introduce in previous commits. Can you squash 6870ee6 and 9e4e0c0 into one commit? I don't like having commits that introduce changes to have them later refined within the same PR. It makes browsing the git log after the PR is merged more difficult.

wezm commented:
Thanks for you patience with me on this. I've more aggressively cleaned up and squashed commits to better stand on their own without subsequent changes. I changed the changelog message and added a FIXME to cmsg structs.

Susurrus commented on src/sys/socket/mod.rs:
```diff
@@ -168,7 +168,7 @@ unsafe fn copy_bytes<'a, 'b, T: ?Sized>(src: &T, dst: &'a mut &'b mut [u8]) {
 
 
-use self::ffi::{cmsghdr, msghdr, type_of_cmsg_len, type_of_cmsg_data};
+use self::ffi::{cmsghdr, msghdr, type_of_cmsg_len, type_of_cmsg_data, type_of_msg_iovlen};
```
Can you alphabetize this list?

Susurrus commented on src/sys/socket/ffi.rs:
```diff
@@ -23,19 +26,28 @@ pub type type_of_cmsg_data = c_uint;
 // actual mutability of data pointed to by this structure.
+//
+// FIXME: Replace these structs with the ones defined in libc
+// E.g. https://doc.rust-lang.org/libc/x86_64-unknown-linux-gnu/libc/struct.msghdr.html
```
You can leave off this link.

Susurrus commented on src/sys/event.rs:
```diff
@@ -79,16 +78,25 @@ pub enum EventFilter {
 }
 
 #[cfg(any(target_os = "macos", target_os = "ios",
```
Please alphabetize the order of the OSes since you're already modifying this attribute.

Susurrus commented on src/sys/event.rs:
```diff
@@ -22,13 +22,12 @@ pub struct KEvent {
-#[cfg(any(target_os = "openbsd", target_os = "freebsd",
-          target_os = "dragonfly", target_os = "macos",
-          target_os = "ios"))]
+#[cfg(any(target_os = "freebsd", target_os = "dragonfly",
```
Please alphabetize the order of the OSes since you're already modifying this attribute.

Susurrus commented on src/sys/signal.rs:
```diff
@@ -488,102 +489,113 @@ pub enum SigevNotify {
+    }
+
+    impl SigEvent {
+        // Note: this constructor does not allow the user to set the
```
Can you turn this into a proper doc comment since you're modifying this? This is definitely something that should be exposed to our end users rather than only visible when browsing the source.

Susurrus commented:
@asomers Can you wrap up your review? I want to make sure all of your concerns are addressed before we merge.

asomers commented:
The CHANGELOG entry is in the `ADDED` section, but that section should only be for new features.  This ia bugfix commit, so its CHANGELOG entry should go into the `FIXED` section.

wezm commented:
All requested changes have been addressed.

wezm commented:
Hi folks, could I trouble you for another review. Would love to get this merged.

asomers commented:
All of my concerns are addressed, and I think susuruss's are too.  I'll merge in 1 day if @Susurrus has no more objections.

Susurrus commented:
bors r+
